Fix #6575 - Datatable : Clean up selection

Rapster opened the pull request:
- Centralize selection code into SelectionFeature
- Keep selection state in view scope, and no longer request scope

No breaking changes

Rapster committed:
Fix #6575 - DataTable: selection gets lost after update only (but not processed)

Rapster committed:
Fix #6575 - DataTable: selection gets lost after update only (but not processed)

Rapster commented on src/main/java/org/primefaces/component/datatable/DataTable.java:
```diff
@@ -294,18 +297,6 @@ public void processValidators(FacesContext context) {
         }
     }
 
-    @Override
-    public void processUpdates(FacesContext context) {
-        super.processUpdates(context);
-
-        ValueExpression selectionVE = getValueExpression(PropertyKeys.selection.toString());
-
-        if (selectionVE != null) {
-            selectionVE.setValue(context.getELContext(), getLocalSelection());
-            setSelection(null);
-        }
-    }
-
```
@tandraschko @melloware I remember you guys faced an issue with this code recently, but I can't recall what it was... Can you remind me what was the issue number?

tandraschko replied:
can only remember that the selection attr was broken in ui:repeat.

Rapster replied:
Got it #3812

tandraschko commented:
cant we get rid of selectedRowKeysAsString? It was easy to remove it for resizable/visible columns at least

Rapster commented:
I'll check and keep you posted

Rapster committed:
Fix #6575 - DataTable: selection gets lost after update only (but not processed)

Rapster committed:
Fix #6575 - DataTable: selection gets lost after update only (but not processed)

Rapster committed:
Fix #6575 - DataTable: selection gets lost after update only (but not processed)

Rapster committed:
Fix #6575 - DataTable: selection gets lost after update only (but not processed)

slepage-klee committed:
Fix #6575 - DataTable: selection gets lost after update only

slepage-klee committed:
Fix #6575 - DataTable: selection gets lost after update only

Rapster committed:
fix merge

Rapster committed:
fix merge

Rapster committed:
Fix #6575 - DataTable: selection gets lost after update only (but not processed)

Rapster committed:
Fix #6575 - DataTable: selection gets lost after update only (but not processed)

Rapster commented:
I think I'm done with this one. Ready for review

tandraschko commented:
did you run the integration tests? Would be great if you could add some more integration tests, so we can be sure that nothing breaks

BTW @melloware / @christophs78 big kudos for the many integration tests! pulled the project today for fixing sorting issues and wow, we really have many of them now.

melloware commented:
The integration tests have come in really handy!  I think we have about 150 tests running nightly now!

melloware commented:
@Rapster we have "single" and "multiple" selection examples already in Integration Tests like DataTable002Test and DataTable010Test so you could probably modify one of those or copy it.

Rapster committed:
Fix #6575 - DataTable: selection gets lost after update only (but not processed)

Rapster committed:
Fix #6575 - DataTable: selection gets lost after update only (but not processed)

Rapster committed:
Fix #6575 - DataTable: selection gets lost after update only (but not processed)

Rapster committed:
Fix #6575 - DataTable: selection gets lost after update only (but not processed)

Rapster committed:
Fix #6575 - DataTable: selection gets lost after update only (but not processed)

Rapster committed:
Fix #6575 - DataTable: selection gets lost after update only (but not processed)

Rapster committed:
Fix #6575 - DataTable: selection gets lost after update only (but not processed)

Rapster committed:
Fix #6575 - DataTable: selection gets lost after update only (but not processed)

Rapster committed:
Fix #6575 - DataTable: selection gets lost after update only (but not processed)

Rapster committed:
Fix #6575 - DataTable: selection gets lost after update only (but not processed)

tandraschko commented:
NVM; is meant saveState :D

tandraschko commented:
the PR looks good; we should just add some integration tests before merge; some that proof selection works like before and 1-2 new ones, which proofs that update works.

tandraschko commented:
if RC1 is released, its too late for such refactorings; i would really like to see this merged before.

tandraschko commented:
and as addRow and editRow was changes, 1-2 simpel integration tests for this feature would be great too.

melloware commented:
@tandraschko We already had an Edit Row test so I just added an Add Row test as well.  

They are in `DataTable007Test` here: https://github.com/primefaces-extensions/primefaces-integration-tests/blob/6d4c8dfb8e7914c77f6b86501f7985c62506284e/src/test/java/org/primefaces/extensions/integrationtests/datatable/DataTable007Test.java

melloware committed:
Merge branch 'master' into fix-6575v2

melloware commented:
I just checked out this PR and built it and ran all the Integration tests including our Single Selection and Multiple Selection, Add Row an Edit Row tests and everything is passing!!!

Rapster committed:
Fix #6575 - DataTable: selection gets lost after update only (but not processed)

Rapster committed:
Merge branch 'fix-6575v2' of https://github.com/Rapster/primefaces into fix-6575v2

Rapster committed:
Fix #6575 - DataTable: selection gets lost after update only (but not processed)

Rapster commented:
Thank you very much @melloware

tandraschko commented on src/main/java/org/primefaces/component/datatable/feature/ScrollFeature.java:
```diff
@@ -82,7 +78,7 @@ public void encode(FacesContext context, DataTableRenderer renderer, DataTable t
     @Override
     public boolean shouldDecode(FacesContext context, DataTable table) {
-        return false;
+        return table.isScrollingRequest(context) && table.isSelectionEnabled();
```
Why did you change all this features?

In the past we always decodde SelectionFeature, just if selection is enabled and indepdent if the current request is a selection request.
So in each request the SelectionFeature was called and decoded if the _selection param is available

Rapster replied:
> Why did you change all this features?

I don't think I change anything, I just adapt according to my changes and move decode stuff in decode method instead of encode method. Arf now that I see it, loading data is also done in encode method... I'll have to change that

> In the past we always decodde SelectionFeature, just if selection is enabled and indepdent if the current request is a selection request.

Maybe, but as of today you have a few features just like this one that requires decoding selection (ex-findSelectedRowKeys). Right now, I just adapt according to my changes, tomorrow everything should probably be done in SelectionFeature

tandraschko replied:
Hmmm thats weird actually.
SelectionFeature#decode runs before all encode, so why does ScrollFeature#encode needs a findSelectedRowKeys call?
Maybe we can remove this calls with your refactoring.

Rapster replied:
I think those calls are to cover default selection, where items has been added right away to the `DataTable#selection` value. For example, you add a new row, and this row should be selected by default, that's what findSelectedRowKeys used to do (and I'm still doing) it's setting new rowKeys through `DataTable#selection` value. But I feel it would be a lot more simple if a new attribute `selectedRow` was added to DataTable, to allow default selection. Item would be then added to `DataTable#selection` once component processed, doing so it would prevent reading `DataTable#selection` at all and figure out what are the rowKeys

tandraschko replied:
That something for V11. I think its better to add this code to DataTableRenderer#preRender or similar isntead of in multiple XXXFeature which are not related to selection.

Rapster replied:
Not that simple, I'd rather come up with a clean solution in V11 than fiddling something now.

Rapster committed:
Fix #6575 - DataTable: clean up selection

Rapster committed:
DataTable: clean up selection

Rapster committed:
DataTable: clean up selection

Rapster committed:
DataTable: clean up selection

Rapster committed:
DataTable: clean up selection

tandraschko commented:
your PR now contains the changes of the other PR which refactored rowKey stuff; im not ok with this for now in this PR.

tandraschko commented:
Its not good to mix such big changes in general as a default rowKey is not related to "cleanup selection" / "save selection in viewstate"
Updates this issue for it: https://github.com/primefaces/primefaces/issues/223

Rapster commented:
From the start it's always been more than just saving selection in view state. This PR is part of the "Clean up series", so yes it involves more changes than usual but, if you look closer it's not that big. SelectionFeature is the one which got a big refactoring.

Regarding the rowKey stuff, that's still part of selection so obviously I modified it a tiny bit (and not as much as my other PR), and again it's a nice to have feature, completely transparent internally and for users. Recently I just aligned the behavior in lazy mode,  as I introduced default hashCode in my very first commits. `LazyDataModel#getRowData` gets a default impl, since ii'm guessing  most of the users (80%?) does the same thing.

#223 Default rowKey and support converter are two different things, so not sure this is the best place to talk about it

If default rowKey really bothers you, I can revert it... and I'll simply introduce it later

tandraschko commented:
`default rowKey` and `converter` are different things but targets the same problem, to default impl this attr/methods.
I think we should avoid the hashCode and the LazyDataModel change or we also work on the Converter thing, that would take your improvement to the next level and is more failsafe as hashCode, if a converter is registered for that object.
HashCode might not be implemented, the Converter IS implemented correctly if available.

That order would be possible:
- use rowKey method/attr if defined
- use converter if available
- fallback to hashCode

Also think the change is quite easy and the converter can be cached as request scoped / field in DataTable.

Rapster commented:
I'm gonna revert default rowKey so this PR gets merged and we can move on. I modify LazyDataModel so it gets cleaner, let me know if you want me to revert it as well

Rapster committed:
DataTable: clean up selection

Rapster committed:
DataTable: clean up selection

tandraschko commented:
ok, looks good to me now

I assume you tested lazy,non-lazy,multi&single-selection, mojarra, myfaces?
then +1

melloware commented:
Our Integration tests test all of the above but only on MyFaces since Tomee only runs Myfaces. So I cant speak for Mojarra.

Rapster commented:
I did all my tests using showcase and mojarra

melloware commented:
OK then lets squash and merge this and I will re-run the Integration Tests on GitHub Actions.
